server/request_handler.go

```diff
@@ -0,0 +1,173 @@
+	// e.g. https://rpc.flashbots.net?url=http://RPC-ENDPOINT.COM
+	customProxyUrl, ok := r.req.URL.Query()["url"]
+	if ok && len(customProxyUrl[0]) > 1 {
+		r.defaultProxyUrl = customProxyUrl[0]
+		r.log("Using custom url: %s", r.defaultProxyUrl)
+	}
+
+	// Decode request JSON RPC
+	defer r.req.Body.Close()
+	body, err := ioutil.ReadAll(r.req.Body)
+	if err != nil {
+		r.logError("failed to read request body: %v", err)
+		r._writeHeaderStatus(http.StatusBadRequest)
+		return
+	}
+
+	if len(body) == 0 {
+		r._writeHeaderStatus(http.StatusBadRequest)
+		return
+	}
+
+	// Parse JSON RPC payload
+	var data interface{}
+	if data, err = r.UnmarshalJSON(body); err != nil {
+		r.log("Parse payload %v", err)
+		r._writeHeaderStatus(http.StatusBadRequest)
+		return
+	}
+
+	// Handle batch request
+	if r.handleBatch {
+		batchCh := make(chan *types.JsonRpcResponse) // Chan to hold response from each go routine
+		wg := new(sync.WaitGroup)                    // Init Wait group
+		for _, req := range data.([]*types.JsonRpcRequest) {
+			wg.Add(1)
+			// Create rpc request
+			rpcReq := NewRpcRequest(r.uid, req, r.handleBatch, body, r.defaultProxyUrl, r.relaySigningKey, origin, ip) // Set each individual request
+
+			// Process each individual request
+			go func(batchWg *sync.WaitGroup, batchReq RequestHandler, ch chan *types.JsonRpcResponse) {
+				defer batchWg.Done()
+				res := batchReq.ProcessRequest()
+				ch <- res
+			}(wg, rpcReq, batchCh)
+		}
+		go func() {
+			// Wait for the batch to get completed
+			wg.Wait()
+			// Close chan
+			close(batchCh)
+		}()
+
+		batchRes := make([]*types.JsonRpcResponse, 0)
+		for res := range batchCh {
+			batchRes = append(batchRes, res) // Add it to batch response list
+		}
+
+		// Write consolidated response
+		r._writeRpcBatchResponse(batchRes)
+		return
+	}
+
+	// Handle single request
+	rpcReq := NewRpcRequest(r.uid, data.(*types.JsonRpcRequest), r.handleBatch, body, r.defaultProxyUrl, r.relaySigningKey, origin, ip)
+	res := rpcReq.ProcessRequest()
+	// Write response
+	r._writeRpcResponse(res)
+}
+
+// UnmarshalJSON implements json.Unmarshaler
+func (r *RpcRequestHandler) UnmarshalJSON(b []byte) (interface{}, error) {
+	if len(b) == 0 {
+		return nil, fmt.Errorf("failed to parse JSON RPC request: no bytes to unmarshal - body: %s", b)
+	}
+	// Based on the first character decide single or multiple
+	switch b[0] {
+	// If payload with single request
+	case '{':
+		var jsonReq *types.JsonRpcRequest
+		if err := json.Unmarshal(b, &jsonReq); err != nil {
+			return nil, fmt.Errorf("failed to parse JSON RPC request: %v - body: %s", err, b)
+		}
+		return jsonReq, nil
+	// If payload with batch request
+	case '[':
+		var jsonBatchReq []*types.JsonRpcRequest
+		if err := json.Unmarshal(b, &jsonBatchReq); err != nil {
+			return nil, fmt.Errorf("failed to parse JSON RPC Batch request: %v - body: %s", err, b)
+		}
+
+		// Set handle batch request
+		r.handleBatch = true
```

Comment:
Maybe we should get rid of `.handleBatch` too and make the Unmarshal function more functional, as in not changing state?

Reply:
Agreed. Removed `handleBatch` from every where